File: artemis-entity-tracker/src/main/java/net/namekdev/entity_tracker/utils/AutoSizedArray.java
package net.namekdev.entity_tracker.utils;

import java.util.ArrayList;

public class AutoSizedArray<T> {
	protected final ArrayList<T> _array;


	public AutoSizedArray() {
		_array = new ArrayList<T>();
	}

	public AutoSizedArray(int capacity) {
		_array = new ArrayList<T>(capacity);
	}

	public T get(int index) {
		return _array.get(index);
	}

	public void set(int index, T value) {
		ensureSize(index+1);
		_array.set(index, value);
	}

	public void ensureSize(int size) {
		_array.ensureCapacity(size);

		int i = _array.size();
		while (i++ < size) {
			_array.add(null);
		}
	}

	public void clear() {
		_array.clear();
	}
}

File: src/commonMain/kotlin/net/namekdev/entity_tracker/utils/AutoSizedArray.kt
package net.namekdev.entity_tracker.utils

open class AutoSizedArray<T> {
    protected val _array: ArrayList<T?>


    constructor() {
        _array = ArrayList()
    }

    constructor(capacity: Int) {
        _array = ArrayList(capacity)
    }

    operator fun get(index: Int): T {
        return _array[index]!!
    }

    operator fun set(index: Int, value: T) {
        ensureSize(index + 1)
        _array[index] = value
    }

    fun ensureSize(size: Int) {
        _array.ensureCapacity(size)

        var i = _array.size
        while (i++ < size) {
            _array.add(null)
        }
    }

    fun clear() {
        _array.clear()
    }
}
